### PR TITLE
Enable eswitch mode setting on SmartNICs

### PR DESCRIPTION
**Pull request by slyon**
## Description
Certain types of network cards/SmartNICs allow toggling their low-level (`devlink`) embedded switch mode PCI setting in order to utilize the full hardware offloading capabilities. With this PR we're exposing this functionality to the netplan YAML schema:
* `embedded-switch-mode: [switchdev|legacy]`
  * choose the devlink setting for the embedded switch (eswitch)
  * sticks to the hardware's preset by default
* `delay-virtual-functions-rebind: [false|true]`
  * False by default: SR-IOV virtual functions are re-bind to their driver right after setting the eswitch mode
  * Can be delayed to a later stage during the boot process; useful when OVS/OVN bonding (VF LAG) is to be used

Example:
```
network:
  version: 2
  ethernets:
    engreen:
      embedded-switch-mode: "switchdev"
      delay-virtual-functions-rebind: true
    enblue:
      match:
        driver: "mlx5_core"
      set-name: enblue
      embedded-switch-mode: "legacy"
      virtual-function-count: 4
    sriov_vf0:
      link: engreen
```

This PR builds upon the new libnetplan YAML parser (https://github.com/canonical/netplan/pull/250) for accessing the new settings.

## Checklist

- [x] Runs `make check` successfully.
- [x] Retains 100% code coverage (`make check-coverage`).
- [x] New/changed keys in YAML format are documented.
- [x] \(Optional\) Adds example YAML for new feature.
- [ ] \(Optional\) Closes an open bug in Launchpad.

### CODE REVIEW

File: netplan/cli/sriov.py
Comment (slyon) [inline netplan/cli/sriov.py:33]: This uses an Apache-2.0 license and can be combined with our GPL3+ code

File: doc/netplan.md
Comment (vorlonofportland) [inline doc/netplan.md:812]: 'eswitch' seems jargony.  Would `embedded-switch-mode` be better?

File: doc/netplan.md
Comment (vorlonofportland) [inline doc/netplan.md:820]: Awkward grammar in this key name.  Even if this is matching a string on the mellanox side, would recommend `delay-virtual-functions-rebind` or `delay-rebinding-virtual-functions` instead.

File: doc/netplan.md
Comment (slyon) [reply doc/netplan.md:812]: Thanks for your comments! Yes, I had thought about that variation as well and will change it accordingly.

**Comment by slyon** [comment]
We got some positive feedback from Canonical's OpenStack team about the new functionality, especially that the VF LAG (bonding of SR-IOV physical functions) is working as expected.

File: netplan/cli/sriov.py
Comment (schopin-pro) [inline netplan/cli/sriov.py:508]: question (non-blocking): given that #254 and #255 are pushed back for later, shouldn't we add tests for those?

This is totally your call, I just want to raise this point since the commit predates the decision.

File: src/sriov.h
Comment (schopin-pro) [inline src/sriov.h:22]: question (blocking, abi): breaking the ABI here used by a binary in `netplan.io`, is this deliberate?

I vaguely recall doing some work in this area, but I don't remember the specifics. More to the point, I don't recall if I tried to maintain ABI with the generate binary or not.

File: src/generate.c
Comment (schopin-pro) [inline src/generate.c:349]: nitpick: the hashtable should be cleared before the state.

There's a brief window where we have a hashtable full of dangling pointers. This isn't a big deal here but is easily avoidable.

File: tests/generator/base.py
Comment (schopin-pro) [inline tests/generator/base.py:472]: suggestion (non-blocking): use the glob to produce the service file

(absolutely untested)
This should take care of the empty map for free, and voids the need to list all the extra service files.

```python
    def assert_sriov(self, file_contents_map):
        systemd_dir = os.path.join(self.workdir.name, 'run', 'systemd', 'system')
        sriov_systemd_dir = glob.glob(os.path.join(systemd_dir, '*netplan-sriov-*.service'))
        self.assertEqual(sriov_systemd_dir, {'netplan-sriov-' + f for f in file_contents_map})
        self.assertEqual(set(os.listdir(self.workdir.name)) - {'lib'}, {'etc', 'run'})
```

File: netplan/cli/sriov.py
Comment (schopin-pro) [inline netplan/cli/sriov.py:413]: thought: potential regressions here (same as #255, really) with newer, stricter checks

**Review by schopin-pro — COMMENTED** [review]
This PR is in great shape, and the code looks overall solid :slightly_smiling_face: 

I have a few items, most of them non-blocking, but there's one in particular regarding potential ABI break that needs answering before merging.

File: tests/test_sriov.py
Comment (slyon) [reply tests/test_sriov.py:123]: yes.

File: netplan/cli/sriov.py
Comment (slyon) [reply netplan/cli/sriov.py:413]: ACK. That was done conscientiously. I had all of the new SR-IOV functionality implemented in a way not using libnetplan parser, but felt it would be better if new functionality would be using the new parser (so no rework is needed later on). A potential regression is limited to the SR-IOV usecase and I could relatively easily change that to using just native python `configmanager.py` in case we hit any issues.

We're also exposing the same potential regression for the new `netplan get` usecase/CLI, which could be mitigated by reverting 331ca01908b68446a9a1f395616136678e617a1a in case of any issue.

But I actually like the idea of enabling the new parser in _some_ selected usecases, to give it some limited exposure before enabling it throughout the stack.

File: src/netplan.c
Comment (slyon) [reply src/netplan.c:711]: ACK.

File: tests/generator/base.py
Comment (slyon) [reply tests/generator/base.py:472]: Nice. Applied with small modifications.

File: netplan/cli/sriov.py
Comment (slyon) [reply netplan/cli/sriov.py:508]: Yes, good point.
* For the first case (`if not vlan_id:`) we do not actually need a test, as this is now just handled a bit earlier by libnetplan, using a slightly different error message; `test_apply_sriov_config_invalid_vlan` is adopted to account for that.
* For the 2nd case (`if not vf:`) I've added a new test case.

File: src/sriov.h
Comment (slyon) [reply src/sriov.h]: Ohh, good spot! I did not do this deliberately, it slipped my mind! We do not want to break the ABI that is used from the `generate` binary (we did that last time, and it was bad).

I've now introduced new API `netplan_state_finish_sriov_write()` and `netplan_sriov_cleanup()` to handle this and moved the old functions to `abi_compat.c`. That should work.

Could you please double check my V2 of commit 9c99162816b6f255173b0adaa906ff9af02b3654 ?

**Comment by slyon** [comment]
Integration tests are still good, even if the CI is doing funny things by skipping most (all?) tests, due to some package dependency issue. I re-ran it locally:
```
autopkgtest [16:07:51]: @@@@@@@@@@@@@@@@@@@@ summary
ovs                  PASS
ethernets            PASS
bridges              PASS
bonds                PASS
routing              PASS
vlans                PASS
wifi                 PASS
tunnels              PASS
scenarios            PASS
regressions          PASS
autostart            PASS
cloud-init           PASS
```

File: netplan/cli/sriov.py
Comment (schopin-pro) [reply netplan/cli/sriov.py:415]: ACK, I'm just happy I'm not the one introducing potential regressions for once :smile:

**Review by schopin-pro — APPROVED** [review]
LGTM!
I have a small suggestion for the new API, but feel free to merge if you don't want to pick it up.

**Comment by slyon** [comment]
Thanks for the re-review! Turned out the `write_sriov_conf_finish()` and `cleanup_sriov_conf()` functions have been part of the `generate` binary itself, so it didn't break ABI in the first place. But still it's nicer to have all this handling of SR-IOV logic in the sriov module and call into the library in future version.

I will be merging this.

**Comment by dshcherb** [comment]
Thanks a lot for this, this is a much needed feature to utilize SR-IOV 👍

In case setting num_vfs needs to be done via systemd in the future, there was some recent work on that too:

https://github.com/systemd/systemd/pull/21865
https://github.com/systemd/systemd/commit/41ce9d769d394d19f1caab8b8a89b7dea50db0bc

**Comment by slyon** [comment]
> In case setting num_vfs needs to be done via systemd in the future, there was some recent work on that too:

Nice! That's good to know. We should probably switch to that upstream solution in the future, once it lands in stable series (maybe in parallel to our manual way of doing it).

**Pull request merged**
